actions/post_actions.jsx

```diff
@@ -31,72 +44,108 @@ const dispatch = store.dispatch;
+                loadNewGMIfNeeded(post.channel_id);
+            }
         }
-    }
+    };
 }
 
-function completePostReceive(post, websocketMessageProps) {
-    if (post.root_id && Selectors.getPost(getState(), post.root_id) == null) {
-        PostActions.getPostThread(post.root_id)(dispatch, getState).then(
-            (data) => {
-                dispatchPostActions(post, websocketMessageProps);
-                PostActions.getProfilesAndStatusesForPosts(data.posts, dispatch, getState);
+export function completePostReceive(post, websocketMessageProps) {
+    return async (doDispatch, doGetState) => {
+        const state = doGetState();
+
+        const rootPost = PostSelectors.getPost(state, post.root_id);
+        if (post.root_id && !rootPost) {
+            const {data: posts} = await doDispatch(PostActions.getPostThread(post.root_id));
+            if (posts) {
+                doDispatch(lastPostActions(post, websocketMessageProps));
             }
-        );
 
-        return;
-    }
+            return;
+        }
 
-    dispatchPostActions(post, websocketMessageProps);
+        doDispatch(lastPostActions(post, websocketMessageProps));
+    };
 }
 
-function dispatchPostActions(post, websocketMessageProps) {
-    const {currentChannelId} = getState().entities.channels;
+export function lastPostActions(post, websocketMessageProps) {
+    return (doDispatch, doGetState) => {
+        const {currentChannelId} = getCurrentChannelId(doGetState());
 
-    if (post.channel_id === currentChannelId) {
-        dispatch({
-            type: ActionTypes.INCREASE_POST_VISIBILITY,
-            data: post.channel_id,
-            amount: 1,
-        });
-    }
+        if (post.channel_id === currentChannelId) {
+            doDispatch({
+                type: ActionTypes.INCREASE_POST_VISIBILITY,
+                data: post.channel_id,
+                amount: 1,
+            });
+        }
 
-    // Need manual dispatch to remove pending post
-    dispatch({
-        type: PostTypes.RECEIVED_POSTS,
-        data: {
-            order: [],
-            posts: {
-                [post.id]: post,
+        // Need manual dispatch to remove pending post
+        doDispatch({
+            type: PostTypes.RECEIVED_POSTS,
+            data: {
+                order: [],
+                posts: {
+                    [post.id]: post,
+                },
             },
-        },
-        channelId: post.channel_id,
-    });
+            channelId: post.channel_id,
+        });
 
-    // Still needed to update unreads
-    AppDispatcher.handleServerAction({
-        type: ActionTypes.RECEIVED_POST,
-        post,
-        websocketMessageProps,
-    });
+        // Still needed to update unreads
+        doDispatch(setChannelReadAndView(post, websocketMessageProps));
 
-    dispatch(sendDesktopNotification(post, websocketMessageProps));
+        doDispatch(sendDesktopNotification(post, websocketMessageProps));
```

Comment:
wondering if you should batch all this into one single dispatch with `batchActions` ? is there a reason not to do that?

Reply:
I tried like below but it didn't worked.  It might be due to dispatching of nested actions.

```
dispatch(batchActions([
    setChannelReadAndView(post, websocketMessageProps),
    sendDesktopNotification(post, websocketMessageProps),
]));
```